src/algorithms/neal2_algorithm.cpp

```diff
@@ -29,7 +29,10 @@ void Neal2Algorithm::print_startup_message() const {
 void Neal2Algorithm::sample_allocations() {
```

Comment:
There are several deliberate changes to this function that appear to alter its functionality -- namely, there is now always a nonzero chance for the label "new cluster" to be drawn, even if the starting cluster is a singleton. Why change it?

Reply:
That's not true. We remove the datum from the cluster as first thing, so that if the datum belongs to a singleton it will have probability zero of being re assigned to it (the weight given by the cardinality is zero)

I changed it this way since it makes code more readable and there are way less copy-paste lines

Reply:
I see that in the case of the starting cluster being a singleton, you're basically moving it at the end of the vector rather than replacing the original one. This is equivalent behaviour so I'm fine with that. You should also change Neal8 then, for code consistency

Reply:
done now